src/bsp/src/stm32/src/USB.cpp

```diff
@@ -0,0 +1,46 @@
+#include "hal/usb.h"
+#include "usbd_cdc_if.h"
+#include <USB.h>
+
+bool USB::send(const uint8_t* buffer, uint16_t length) {
+    if (buffer == nullptr || length > CBUFF_USB_DATA_SIZE) {
+        m_logger.log(LogLevel::Warn, "Invalid parameters for USB::send");
+        return false;
+    }
+
+    bool ret = false;
+    int out = Usb_Send_Data(const_cast<uint8_t*>(buffer), length);
+
+    if(out == USBD_OK) {
+        ret = true;
+    }else{
+        m_logger.log(LogLevel::Warn, "USB_Send_Data was not able to send the data");
+    }
+
+    return ret;
+}
+
+bool USB::receive(uint8_t* buffer, uint16_t length) {
+    if (buffer == nullptr || length > CBUFF_USB_DATA_SIZE) {
+        m_logger.log(LogLevel::Warn, "Invalid parameters for USB::Receive");
+        return false;
+    }
+
+    m_receivingTaskHandle = xTaskGetCurrentTaskHandle();
+    while (CircularBuff_getLength(&cbuffUsb) < length) {
+        // Gets notified everytime a new packet is appended to cbuffUsb
+        ulTaskNotifyTake(pdTRUE, 500);
```

Comment:
You aren't notifying this anywhere, so you end up just doing a sleep of 500ms. Either change this for a sleep (not ideal for responsiveness) or have `Usb_CDC_RxCallBack()` notify this or call a C++ callback which does.

Reply:
notification has been addressed